lib/awful/keygrabber.lua

```diff
@@ -45,7 +207,411 @@ function keygrabber.stop(g)
+-- on the same key combination, they are all executed and there is almost no
+-- safe way to undo that. Also, while `is_enabled` can be used to temporarely
+-- turn off this behavior, it currently cannot be detached completely. Make sure
+-- the `keygrabber` that use this option have a single instance.
+--
+-- @property export_keybindings
+-- @param[opt=false] boolean
+
+--- The root (global) keybinding to start this keygrabber.
+--
+-- Instead of adding an entry to `root.keys` or `rc.lua` `globalkeys` section,
+-- this property can be used to take care of everything. This way, it becomes
+-- easier to package modules using keygrabbers.
+--
+-- @DOC_text_awful_keygrabber_root_keybindings_EXAMPLE@
+--
+-- @property root_keybindings
+-- @see export_keybindings
+-- @see keybindings
+
+--- The keybindings associated with this keygrabber.
+--
+-- The keybindings syntax is the same as `awful.key` or `awful.prompt.hooks`. It
+-- consists of a table with 4 entries.
+--
+-- * `mods` A table with modifier keys, such as `shift`, `mod4`, `mod1` (alt) or
+--  `control`.
+-- * `key` The key name, such as `left` or `f`
+-- * `callback` A function that will be called when the key combination is
+--  pressed.
+-- * `description` A table various metadata to be used for `awful.hotkeys_popup`.
+--
+-- @property keybindings
+-- @param table
+-- @see export_keybindings
+-- @see root_keybindings
+
+--- Define a list of keys that are allowed instead of using `release_key`.
+--
+-- When defined, this property allows to define an implicit way to release the
+-- keygrabber. It helps save some boilerplate code in the handler callbacks.
+--
+-- @DOC_text_awful_keygrabber_allowed_keys_EXAMPLE@
+--
+-- @property allowed_keys
+-- @tparam[opt=nil] table|nil allowed_keys The list of keys.
+
+--- The sequence recorded from the current acquisition.
+--
+-- In this example, the `stop_callback` is used to retrieve the final key
+-- sequence.
+--
+-- Please note that backspace will remove elements from the sequence and
+-- named keys and modifiers are ignored.
+--
+-- @DOC_text_awful_keygrabber_autostart_EXAMPLE@
+--
+-- @property sequence
+-- @param string
+--
+
+--- The current (running) instance of this keygrabber.
+--
+-- The keygrabber instance is created when the keygrabber starts. It is an object
+-- mirroring this keygrabber object, but where extra properties can be added. It
+-- is useful to keep some contextual data part of the current transaction without
+-- poluting the original object of having extra boilerplate code.
+--
+-- @tfield keygrabber current_instance
+-- @see property::current_instance
+
+--- The global signal used to track the `current_instance`.
+--
+-- @signal property::current_instance
+
+--- If a keygrabber is currently running.
+-- @tfield boolean is_running
+
+--- Start the keygrabber.
+--
+-- Note that only a single keygrabber can be started at any one time. If another
+-- keygrabber (or this one) is currently running. This method returns false.
```

Comment:
I thought one of the goals was to make it possible to stack keygrabbers and have multiple currently running ones?!?
(But supporting that might not actually be a good idea...)